src/index/base.cpp

```diff
@@ -140,7 +140,21 @@ bool BaseIndex::Init()
     // Note: this will latch to true immediately if the user starts up with an empty
     // datadir and an index enabled. If this is the case, indexation will happen solely
     // via `BlockConnected` signals until, possibly, the next restart.
-    m_synced = start_block == active_chain.Tip();
+    if (start_block == active_chain.Tip()) {
+        // To prevent race conditions, m_synced = true needs to be set from the
+        // validationinterface thread and the m_synced = true callback needs to
+        // be queued while cs_main is held.
+        //
+        // To prevent older, stale notifications currently in the validation
+        // queue from being processed, it is important setting m_synced = true
+        // in queue after they processed.
+        //
+        // To prevent new notifications that may be happening in the background
+        // from being lost, it is important to keep cs_main locked while calling
+        // CallFunctionInValidationInterfaceQueue, so the new notifications will
+        // be queued after the m_synced = true callback.
+        CallFunctionInValidationInterfaceQueue([this] { m_synced = true; });
+    }
```

Comment:
In ba30b60b:

Isn't this conflicting with the `StartBackgroundSync()` call?

Let's say that there are new blocks on the notifications queue at this point. By not setting the `m_synced` flag, we allow `ThreadSync()` to start.
This former one might have some work to do if some of the queued blocks notifications were processed in-between `Init()` and `StartBackgroundSync()`. Then, while this is being done, the validation queue thread might end up setting `m_synced=true`. Which would enable the index new block connection signals reception at the same time that `ThreadSync()` is being executed.